### ISSUE TITLE
Difficult to integrate

### ISSUE THREAD

**Issue by mrifni** [readium/r2-testapp-kotlin#386]
I am finding it a bit difficult to integrate readium SDK in my app using the "r2-testapp-kotlin" sample.
Shouldn't it be simple like `SomeClass.openEpub(filePath, arguments:{})` ? 
I can see a server needs to start ( whatttt?? a server inside a mobile app ? ) as well?
I am bit lost, is there any integration guide available apart from the sample app.

cheers guys, thank you for the awesome job.

**Comment by mickael-menu** [comment]
Yes, we are aware that the project is lacking proper documentation. It's something we plan to improve in the short term.

But in the meantime, you can take a look at https://readium.org/architecture/ which explains how each Readium component fits into the bigger picture.

> Shouldn't it be simple like `SomeClass.openEpub(filePath, arguments:{})` ? 

Of course we strive to make the APIs simple to use, but Readium is not really a "plug-and-play" library. The ebook world is quite complex and we want to offer performant and flexible components to build advanced reading apps. In particular, Readium supports many other formats and not only EPUB. Not to say that we don't want to offer high-level APIs, but it will never be as simple as `Readium.openEPUB(file)`.

> I can see a server needs to start ( whatttt?? a server inside a mobile app ? ) as well?

That's one of the pain point is the current API. We need an HTTP server because an EPUB is actually a bunch of HTML, CSS and JavaScript files packaged together. Up until now, using an HTTP server was actually the simplest and most reliable way to render an EPUB 3 file.

However, we want to explore [some alternative](https://github.com/readium/r2-testapp-kotlin/issues/297) this year, which could simplify things. But we don't have any guarantee that it will work properly yet. In the past we had issue rendering embedded video files with such solution.

## How to integrate Readium 2 in your app?

I'll take a shot at an overview on how to integrate Readium Mobile Kotlin in your app. Note that this assume that you are using at least 2.0.0-beta.1.

### 1. Parse a `Publication` from an EPUB

First, you will need to parse an EPUB into a high-level `Publication` object using the Streamer.

```kotlin
val streamer = Streamer(context)

val asset = FileAsset(File(filePath))
val publication = streamer.open(asset, allowUserInteraction = true, sender = activity)
    .getOrElse { error ->
        showErrorMessage(error.getUserMessage(context))
        return
    }

// This is checking for publications protected by DRM
if (publication.isRestricted) {
    publication.protectionError?.let { error ->
        showErrorMessage(error.getUserMessage(context))
    }
    return
}

// Now you can use the publication...
```

From there, you can use the `Publication` object to get the metadata for this EPUB file, for example to add it to a user library database.

But if you want to display the EPUB content to the user, continue with the next sections.

### 2. Add the `Publication` to the HTTP server

This is necessary to display the EPUB to the user with a Navigator, as explained earlier.

The API is really not great if you use the HTTP server with `R2EpubActivity`, so I'll assume that you will use your own `Activity` and the `EpubNavigatorFragment` instead. In which case you can do something like:

```kotlin
val server: Server by lazy {
    val s = ServerSocket(if (DEBUG) 8080 else 0)
    s.localPort
    s.close()
    return Server(s.localPort, context)
}

fun serve(publication: Publication, asset: PublicationAsset) {
    // This is the path where the user settings will be saved, you need to use the same path that
    // you use in `UserSettings.kt` (taken from the testapp). By default it looks like this:
    val userProperties = File(File(context.filesDir.path, Injectable.Style.rawValue), "UserProperties.json").path
    server.addEpub(publication, null, "/${asset.name}", userProperties)
}
```

### 3. Present the `Publication` with a Navigator

After adding the publication to the HTTP server, you are ready to show the reading activity. You can check what kind of ebook it is with `asset.mediaType()`.

```kotlin
val intent = when (asset.mediaType()) {
    MediaType.EPUB -> Intent(context, EpubActivity::class.java).apply {
        ...
    }
    else -> null
}

if (intent != null) {
    context.startActivity(intent)
}
```

In your own `EpubActivity`, you need to inflate the `EpubNavigatorFragment` using the factory provided by Readium:

```kotlin
override fun onCreate(savedInstanceState: Bundle?) {
    // Needs to be called before super.
    supportFragmentManager.fragmentFactory = EpubNavigatorFragment.createFactory(
        publication = publication,
        // To get the URL where the publication is served from the server, you need to pass
        // the publication filename and the server port to this static function: 
        baseUrl = Publication.localBaseUrlOf(asset.name, server.port)
    )

    super.onCreate(savedInstanceState)
    
    setContentView(R.layout.activity_r2_epub)
}
```

Handling the base URL really suck at the moment, I'll take a look to add a `server.addPublication()` API which will return the base URL directly instead in the next beta.

**Comment by mrifni** [comment]
thank you for the detailed explanation

**Comment by mickael-menu** [comment]
I fixed the Server API in:
* https://github.com/readium/r2-streamer-kotlin/pull/142
* https://github.com/readium/r2-navigator-kotlin/pull/196
* https://github.com/readium/r2-testapp-kotlin/pull/387

This will be available in the next beta version released probably this week.

The difference is that you don't need to give the publication filename when adding the publication to the server, and it will return the base URL to give to the navigator fragment or activity:

```kotlin
fun serve(publication: Publication, asset: PublicationAsset): URL? {
    // This is the path where the user settings will be saved, you need to use the same path that
    // you use in `UserSettings.kt` (taken from the testapp). By default it looks like this:
    val userProperties = File(context.filesDir, Injectable.Style.rawValue + "/UserProperties.json")
    return server.addPublication(publication, userPropertiesFile)
}
```

This makes it easier to use directly `R2EpubActivity` too, which requires a new `baseUrl` intent extra.

**Comment by mrifni** [comment]
cheers

**Comment by mrifni** [comment]
can we merge the PRs to develop branches ? So I can test it out

**Comment by mickael-menu** [comment]
Done. If you want to use `R2EpubActivity` which is more "plug-and-play", take a look at:

https://github.com/readium/r2-testapp-kotlin/blob/develop/r2-testapp/src/main/java/org/readium/r2/testapp/utils/NavigatorContract.kt

and

https://github.com/readium/r2-testapp-kotlin/blob/6693ade584b5bfe541f69f6b8775d2a982dd399b/r2-testapp/src/main/java/org/readium/r2/testapp/library/LibraryActivity.kt#L581-L608

**Comment by mrifni** [comment]
ive got it working now, thanks
